### PR TITLE
Remove substation modification

### PR DESCRIPTION
**Pull request by souissimai**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

**What is the current behavior?**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

### CODE REVIEW

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package com.powsybl.iidm.modification.topology;
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:1]: Licence

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.powsybl.iidm.modification.topology;
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstationBuilder.java:1]: Licence

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * @author Maissa Souissi <maissa.souissi at rte-france.com> eurostag-remove-voltage-level-bb.xml
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:29]: eurostag-remove-voltage-level-bb.xml to remove

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/util/ModificationReports.java [hunk 4]
```diff
@@ -169,6 +170,15 @@ public static void removeFeederBayAborted(Reporter reporter, String connectableI
                 .build());
     }
 
+    public static void removedSubstationReport(Reporter reporter, String substationId) {
+        reporter.report(Report.builder()
+                .withKey("removeSubstation")
+                .withDefaultMessage("Substation ${substationId}, its voltage levels have been removed")
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/util/ModificationReports.java:176]: Substation ${substationId} and its voltage levels have been removed

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            }
+            return;
+        }
+        List<VoltageLevel> voltageLevels = new ArrayList<>(substation.getVoltageLevelStream().collect(Collectors.toList()));
+        for (VoltageLevel voltageLevel : voltageLevels) {
+            new RemoveVoltageLevel(voltageLevel.getId()).apply(network, true, reporter);
+        }
+        substation.remove();
+        removedSubstationReport(reporter, substationId);
+        LOGGER.info("Substation {}, and its voltage levels have been removed", substationId);
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:51]: Substation {} and its voltage levels have been removed

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static com.powsybl.iidm.modification.util.ModificationReports.notFoundSubstationReport;
+import static com.powsybl.iidm.modification.util.ModificationReports.removedSubstationReport;
+
+/**
+ * @author Maissa Souissi <maissa.souissi
+ * at rte-france.com>
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:23]: On one row

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        assertEquals(Set.of("GEN", "NHV1", "P1", "VLHV1", "VLGEN", "NGEN_NHV1", "NGEN", "NHV1_NHV2_2", "NHV1_NHV2_1"), beforeRemovalObjects);
+        assertEquals(Set.of("GEN", "NHV1", "P1", "VLHV1", "VLGEN", "NGEN_NHV1", "NGEN", "NHV1_NHV2_2", "NHV1_NHV2_1"), removedObjects);
+        assertNull(network.getSubstation("P1"));
+        assertNull(network.getVoltageLevel("VLGEN"));
+        assertNull(network.getVoltageLevel("VLHV1"));
+        assertNull(network.getTwoWindingsTransformer("NGEN_NHV1"));
+    }
+
+    @Test
+    public void testRemoveSubstation() {
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:82]: You should remove the public

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        Substation substation = network.getSubstation(substationId);
+        if (substation == null) {
+            LOGGER.error("Substation {} not found", substationId);
+            notFoundSubstationReport(reporter, substationId);
+            if (throwException) {
+                throw new PowsyblException("Substation not found: " + substationId);
+            }
+            return;
+        }
+        List<VoltageLevel> voltageLevels = new ArrayList<>(substation.getVoltageLevelStream().collect(Collectors.toList()));
```
Comment (etiennehomer) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:45]: You can make it simpler : 
List<VoltageLevel> voltageLevels = substation.getVoltageLevelStream().collect(Collectors.toList());

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        for (VoltageLevel voltageLevel : voltageLevels) {
+            new RemoveVoltageLevel(voltageLevel.getId()).apply(network, true, reporter);
+        }
```
Comment (SlimaneAmar) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:48]: Replace by 
`       substation.getVoltageLevelStream().map(VoltageLevel::getId).forEach(id -> new RemoveVoltageLevel(id).apply(network, true, reporter));`

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * @author Maissa Souissi <maissa.souissi
+ * at rte-france.com>
+ */
+public class RemoveSubstation extends AbstractNetworkModification {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RemoveSubstation.class);
+
+    private final String substationId;
+
+    public RemoveSubstation(String substationId) {
```
Comment (SlimaneAmar) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:30]: Remove public

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+class RemoveSubstationTest extends AbstractConverterTest {
+    private final Set<String> removedObjects = new HashSet<>();
+    private final Set<String> beforeRemovalObjects = new HashSet<>();
+
+    @AfterEach
+    public void tearDown() {
+        removedObjects.clear();
+    }
+
+    private void addListener(Network network) {
```
Comment (SlimaneAmar) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:40]: Why we use a `beforeRemovalObjects` ?
It'is always the same as `afterRemoval` ?

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            }
+
+            @Override
+            public void afterRemoval(String id) {
+                removedObjects.add(id);
+            }
+        });
+    }
+
+    @Test
```
Comment (SlimaneAmar) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:54]: I don't undersatnd the diffrence between the 3 tests

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        assertNull(network.getVoltageLevel("VLHV1"));
+        assertNull(network.getTwoWindingsTransformer("NGEN_NHV1"));
+    }
+
+    @Test
+    public void testRemoveSubstation() {
+        Network network = FourSubstationsNodeBreakerFactory.create();
+        addListener(network);
+        assertNotNull(network.getSubstation("S1"));
+        assertNotNull(network.getSubstation("S1").getVoltageLevels());
```
[the remark this answers is not in the document]
Comment (souissimai) [reply iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:86]: I added those line to confirm that the substation exists and it has voltage levels before removal

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 8 above)
Comment (SlimaneAmar) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:45]: List<String> vlIds = substation.getVoltageLevelStream().map(VoltageLevel::getId).collect(Collectors.toList());
        vlIds.forEach(id -> new RemoveVoltageLevel(id).apply(network, true, reporter));

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 9 above)
Comment (SlimaneAmar) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:48]: List<String> vlIds = substation.getVoltageLevelStream().map(VoltageLevel::getId).collect(Collectors.toList());
        vlIds.forEach(id -> new RemoveVoltageLevel(id).apply(network, true, reporter));

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+    @AfterEach
+    public void tearDown() {
+        removedObjects.clear();
+    }
+
+    private void addListener(Network network) {
+        network.addListener(new DefaultNetworkListener() {
+            @Override
+            public void beforeRemoval(Identifiable id) {
```
Comment (SlimaneAmar) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:43]: Remove this list

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 10 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:35]: Why do you want to remove the public ? 
We have a public constructor for all Remove modifications.

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java (hunk 3 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java:25]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 1 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:8]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstationBuilder.java (hunk 2 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstationBuilder.java:8]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/util/ModificationReports.java (hunk 4 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/util/ModificationReports.java]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 5 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 6 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java]: done

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java (hunk 7 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java]: done

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Test
+    void testRemoveSubstationWithManyVL() throws IOException {
+        Network network = EurostagTutorialExample1Factory.create();
+        addListener(network);
+        assertEquals("VLGEN", network.getVoltageLevel("VLGEN").getId());
+        assertEquals("VLHV1", network.getVoltageLevel("VLHV1").getId());
+        new RemoveSubstationBuilder().withSubstationId("P1").build().apply(network);
+        System.out.println(network.getVoltageLevel("VLGEN"));
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/RemoveSubstationTest.java]: done

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 8 above)
Comment (etiennehomer) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java]: done

**Review by etiennehomer — APPROVED** [review]

**Review by rolnico — APPROVED** [review]
Constructor for Remove modifications could also be set to default as are all the Create modifications (in order to force the user to use the builder) but that change can be done later.

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java (hunk 10 above)
Comment (rolnico) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/RemoveSubstation.java:35]: A new PR will be opened in order to modify the accessors and make them coherent in all the modifications (remove and create)

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**